src/wasm-io.h

```diff
@@ -40,8 +40,13 @@ class ModuleReader : public ModuleIO {
   void readText(std::string filename, Module& wasm);
   // read binary
   void readBinary(std::string filename, Module& wasm);
+  // read object file
```

Comment:
The only issue I'm aware of is the use of mutable global imports. Are there others?

One way to fix mutable global imports is to have Binaryen support the proposed [mutable global import](https://github.com/WebAssembly/threads/blob/master/proposals/threads/Globals.md) feature.

Reply:
I thought it was a strict goal to have wasm object files be proper wasm files, plus some custom segments? Them not validating seems surprising, is that documented in the conventions repo?

Or is mutable global imports a "import from future" feature in the python sense? that's how I understand you @sunfishcode ?

Reply:
(To be clear, my motivation here is that I really hope we don't need a new "wasm object file" mode in binaryen.)

Reply:
Mutable global imports is an "import from future" feature, where the future is the threads proposal.

Reply:
I see, thanks. Then I think we should allow mutable global imports instead of adding a mode for object files, unless there is some other reason we need that (I hope not!).

Reply:
I think Clang/LLD currently always emit the stack pointer as a mutable global - but it's not exported since that would be illegal.

Do you use `--relocatable/-r` or just `--emit-relocs`  when you're running LLD? `--relocatable` makes LLD output a merged "object file" rather than the final Wasm output, and `--emit-relocs` is a sort of halfway house, where it emits a "final" Wasm output but with the linking metadata. It's a bit broken (at least on this stack pointer issue) because `--emit-relocs` on its own will create the stack pointer global, yet without exporting it, so relocations can't really be processed by the next stage :(

I'm asking because I have an LLD review open at the moment that explicitly breaks `--emit-relocs`. I'm thinking though that for your purposes, it's probably what you want, and I should instead fix `--emit-relocs` to export the SP as a mutable global. And finally, `wasm-emscripten-finalize` would strip the stack-pointer export.

Reply:
The need to read object files is for `wasm-link-metadata`, which reads a wasm object file pre-LLD. We don't use `--relocatable` or `--emit-relocs`, and we don't care if the stack pointer is exported or not (`wasm-emscripten-finalize` adds stack-manipulating functions as part of its interface with Emscripten).

So, no problems

Reply:
FYI, importing+exporting mutable globals has been separated out into its own proposal: https://github.com/WebAssembly/mutable-global.

Reply:
Thanks - I've been looking forward to mutable global exports landing in browsers! Then you can get rid of some of your helper functions maybe too, once js can manipulate the stack.